modelling/src/main/java/org/axonframework/modelling/StateManager.java

```diff
@@ -29,15 +29,50 @@
  * persist them. Implementations may specify whether they load entities through
  * {@link Repository#load(Object, ProcessingContext)} or {@link Repository#loadOrCreate(Object, ProcessingContext)}.
  * <p>
- * Entities are registered by their type in combination with their id. In case an id is given that matches two
- * repositories, the repository that is registered for the most specific type is used. If no repository is found for the
- * given type and id, a {@link MissingRepositoryException} is thrown.
+ * Entities are {@link #register(Repository) registered} by their type in combination with their id. The combination of
+ * {@link Repository#entityType() entity type} and {@link Repository#idType() id type} of all repositories must be
+ * unique and unambiguous. This means you cannot register a repository if another conflicting repository already exists.
+ * If you do, a {@link ConflictingRepositoryAlreadyRegisteredException} will be thrown. Note that superclasses and
+ * subclasses of each other are considered conflicting.
  *
  * @author Mitchell Herrijgers
  * @since 5.0.0
  */
 public interface StateManager {
 
+    /**
+     * Registers an {@link Repository} for use with this {@code StateManager}. The combination of
+     * {@link Repository#entityType()} and {@link Repository#idType()} must be unique for all registered repositories.
+     * If a repository with the same combination is already registered, a
+     * {@link ConflictingRepositoryAlreadyRegisteredException} is thrown.
+     *
+     * @param repository The {@link Repository} to use for loading state.
+     * @param <I>        The type of id.
+     * @param <T>        The type of the entity.
+     * @return This {@code StateManager} for fluent interfacing.
+     */
+    <I, T> StateManager register(Repository<I, T> repository);
```

Comment:
Nit: Would be good to add some `@Nonnull` annotations here :-)

Reply:
Added, as well as nullchecks in implementations